README.md
# Lichess-Board
This program is used to send moves from a local board to lichess trough the api.


changelog
1.0.0.0
Grundversion ist fertig

1.0.1.0
async-lichess-sdk Libary entfernt, durch reqeust.post ersetzt

1.0.2.0
absofort wird überprüft, ob sie(die Figur) aufgehoben/abgesetzt wurde

1.0.2.1
Fehlerbehebung

1.0.3.0
keine zwei Seriellle Daten erforderlich, wurde durch Arduino gelöst
max. Anzahl an Änderungen pro Datensatz

1.0.4.0
2 Daten weider erforderlich

1.0.4.1
Fehlerbehebung
cut out Seriell nutzlose Daten

1.0.4.2
Fehlerbehebung

1.0.5.0
Code kompremierung
Unnötiges Zeug entfernt

1.0.6.0
requirments mit pip installieren

1.0.6.1
Unbekannte Playing Id, konnte nicht ausgelsen wurde
Code kompremierung, unnötige Befehle entfernt

1.0.6.2
Raiset Exception wenn kein Spiel vorhanden ist --> Fixed
requirments.txt angepasst --> added one Libary

1.0.7.0
Animation, bei Board nicht connected

zug_test.py
import requests
import asyncio
from lichess_client import APIClient
import serial
import time

user= "lucakreativ"
meintoken=""

mcount=0
halb_moves=["", ""]
#id_game="VKdMtBsg"
#moves="e2e4"

felder_name=[
"a1", "b1", "c1", "d1", "e1", "f1", "g1", "h1",
"a2", "b2", "c2", "d2", "e2", "f2", "g2", "h2",
"a3", "b3", "c3", "d3", "e3", "f3", "g3", "h3",
"a4", "b4", "c4", "d4", "e4", "f4", "g4", "h4",
"a5", "b5", "c5", "d5", "e5", "f5", "g5", "h5",
"a6", "b6", "c6", "d6", "e6", "f6", "g6", "h6",
"a7", "b7", "c7", "d7", "e7", "f7", "g7", "h7",
"a8", "b8", "c8", "d8", "e8", "f8", "g8", "h8",
]

altser_data=["x","x","x"]
filter_data=["x","x"]


lesestat=0
#time.sleep(3)

def Serial():
    #print(mcount)
    try:
        ser = serial.Serial('/dev/ttyUSB1', 9600)

    except serial.serialutil.SerialException:
        print("Bitte DGT-Brett anschließen")
        #lesSerial()

    else:
        #print("angeschlossen")
        try:
            ser_data = ser.readline()
        except serial.serialutil.SerialException:
            print("Bitte DGT-Brett anschließen")
        else:
            #print(ser_data)
            if len(ser_data)>=66:
                #print("Groß Genug")

                altser_data[2]=altser_data[1]
                altser_data[1]=altser_data[0]
                altser_data[0]=ser_data

                if altser_data[2]!= "x":
                    #print("nicht x")
                    print(ser_data)
                    
                    if altser_data[0]!=altser_data[1]:
                        Serial()
                    else:
                        filter_data[1]=filter_data[0]
                        filter_data[0]=altser_data[0]

                        if filter_data[1]!="x":
                            Felder(filter_data[0], filter_data[1], mcount)
                else:
                    print("Keine Referenz")
            else:
                print("Datenübertragung ist zu kurz")


def checkGame():
    userdatajason= requests.get("https://lichess.org/api/user/"+user)
    userdata=userdatajason.text

    posP = userdata.find("playing")

    if posP<=1250:
        #print("Spielt gerade")

        id_game=userdata[posP+30:posP+38]
        #print(id_game)
        return(id_game)
    else:
        #print("Spielt gerade nicht")
        return("0")


def Felder(ser_data, altser_datan, mcount):
    if ser_data!=altser_datan:
        #print("neues Spielfeld")
        #print(altser_data)
        #print(halb_moves[0])

        #print(mcount)

        #i=0
        for i in range(64):
            if ser_data[i]!=altser_datan[i]:
                if halb_moves[0]=="":
                    ms=felder_name[i]
                    print("Erst: "+ms)

                    halb_moves[1]=halb_moves[0]
                    halb_moves[0]=ms

                    #if halb_moves[1]==halb_moves[0]:
                        #mcount=0
                    #else:
                    mcount=1


                elif halb_moves!="":
                    me=felder_name[i]
                    print("Zweit :"+me)

                    halb_moves[1]=halb_moves[0]
                    halb_moves[0]=me

                    #if halb_moves[0]==halb_moves[1]:
                        #mcount=1
                    #else:
                    move=halb_moves[1]+halb_moves[0]

                    halb_moves[0]=""
                    halb_moves[1]=""


                    mcount=0
                    Move(move)


        i+=1


def Move(move):
    print()
    print("Zug ist: "+move)
    sendMove(move)
    """
    rf = input("Richtig/Fasch (y/n): ")
    if rf=="y":
        sendMove(move)
        print()
    elif rf=="n":
        print()
        print("Zürückstellen innerhalb 5 Sekunden und dann nochmal machen")
        time.sleep(5)
    else:
        print()
        print("Bitte Ja oder Nein eingeben (y/n)")
        Move(move)
    """

def sendMove(move):
    print()

    game_id=checkGame()
    if game_id!="0":
        loop = asyncio.get_event_loop()
        loop.run_until_complete(main(move, game_id))
    else:
        print("Du bist gerade in keinem Spiel, bitte Figur zurückstellen(innerhalb 5 Sekunden) und ein Spiel starten")


async def main(move, id_game):
    client = APIClient(token=meintoken)
    response = await client.boards.make_move(game_id=id_game, move=move)
    #print(response)
    Lichess_response(response)


def Lichess_response(response):
    print(response)
    """pos=find.response("content")
    posE=response.find("}", pos)
    zurück_lichess=response[pos+12:posE-1]
    pos=zurück_lichess.find("ok")

    if pos==-1:
        print(zurück_lichess)
    else:
        print("Zug wurde erfolgreich ausgeführt")"""




while True:
    #lesestat=Serial(lesestat)
    Serial()
    
    time.sleep(0.1)

main.py
import time
import json
from requests.models import Response
import os

import serial
import requests


user= ""
meintoken=""

mcount=0
halb_moves=["", ""]
backs_moves=["", ""]
USB_Port="0"

#0 for nicht senden und 1 für senden
senden_ja_nein=1
animation=["|","/","-","\\", "|", "-", "\\"]


anima_count=[0]


felder_name=[
"a1", "b1", "c1", "d1", "e1", "f1", "g1", "h1",
"a2", "b2", "c2", "d2", "e2", "f2", "g2", "h2",
"a3", "b3", "c3", "d3", "e3", "f3", "g3", "h3",
"a4", "b4", "c4", "d4", "e4", "f4", "g4", "h4",
"a5", "b5", "c5", "d5", "e5", "f5", "g5", "h5",
"a6", "b6", "c6", "d6", "e6", "f6", "g6", "h6",
"a7", "b7", "c7", "d7", "e7", "f7", "g7", "h7",
"a8", "b8", "c8", "d8", "e8", "f8", "g8", "h8",
"z"
]

altser_data=["x","x","x"]
filter_data=["x","x"]


lesestat=0
#time.sleep(3)

def Serial(anima_count):
    #print(mcount)
    try:
        ser = serial.Serial('/dev/ttyUSB'+USB_Port, 9600)

    except serial.serialutil.SerialException:
        os.system('clear')
        print("Bitte DGT-Brett anschließen ("+animation[anima_count[0]]+")")
        anima_count[0]+=1
        if anima_count[0]>=7:
            anima_count[0]=0
        time.sleep(0.5)

        

    else:
        try:
            ser_data = ser.readline()
        except serial.serialutil.SerialException:
            print("Bitte DGT-Brett anschließen")
        else:
            ser_data=str(ser_data)
            startd=ser_data.find("s//")+3
            endd=ser_data.find("//s")
            ser_data=ser_data[startd:endd]


            if len(ser_data)>=65:
                altser_data[2]=altser_data[1]
                altser_data[1]=altser_data[0]
                altser_data[0]=ser_data

                if altser_data[2]!= "x":

                    filter_data[1]=filter_data[0]
                    filter_data[0]=altser_data[0]

                    if filter_data[1]!="x":
                        maxChanges(filter_data[0], filter_data[1], mcount)
                else:
                    print("Keine Referenz")
            else:
                print("Datenübertragung ist zu kurz")


def checkGame():

    userdatajason= requests.get("https://lichess.org/api/user/"+user)
    userdata=userdatajason.text
    json_user = json.loads(userdata)

    try:
        url=json_user["playing"][20:28]
    except KeyError:
        print("You arent Playing a Game right now")
        return "0"
    else:
        return url


def maxChanges(ser_data, altser_datan, mcount):
    changesz=0
    if ser_data!=altser_datan:
        for i in range(65):
            if ser_data[i]!=altser_datan[i]:
                changesz+=1
    Felder(ser_data, altser_datan, mcount, changesz)


def Felder(ser_data, altser_datan, mcount, changesz):
    if changesz<=2:

        for i in range(65):
            if ser_data[i]!=altser_datan[i]:
                if ser_data[i]=="0":
                    if halb_moves[0]=="":
                        ms=felder_name[i]

                        if ms!="z":
                            print("Erst: "+ms)
                                                      
                        halb_moves[0]=ms


                elif ser_data[i]=="1":
                    if halb_moves[0]!="":
                        me=felder_name[i]

                        print("Zweit :"+me)

                        halb_moves[1]=halb_moves[0]
                        halb_moves[0]=me

                        backs_moves[1]=halb_moves[1]
                        backs_moves[0]=halb_moves[0]

                        move=halb_moves[1]+halb_moves[0]

                        halb_moves[1]=halb_moves[0]
                        halb_moves[0]=""

                        if halb_moves[1]!="z" and halb_moves[0] != "z":
                            Move(move)
                        else:
                            print("Halbzug Manuell eingegeben")


        i+=1


def Move(move):
    print()
    print("Zug ist: "+move)
    if senden_ja_nein==1:
        sendMove(move)
    

def sendMove(move):
    print()

    game_id=checkGame()
    if game_id!="0":
        resoponse=requests.post("https://lichess.org/api/board/game/"+game_id+"/move/"+move, headers={"Authorization":"Bearer "+meintoken})
        print(resoponse.text)


while True:
    Serial(anima_count=anima_count)

zensiert.py
import requests
import asyncio
from lichess_client import APIClient
import serial
import time

user= "lucakreativ"
meintoken=""

mcount=0
halb_moves=["", ""]
#id_game="VKdMtBsg"
#moves="e2e4"

felder_name=[
"a1", "b1", "c1", "d1", "e1", "f1", "g1", "h1",
"a2", "b2", "c2", "d2", "e2", "f2", "g2", "h2",
"a3", "b3", "c3", "d3", "e3", "f3", "g3", "h3",
"a4", "b4", "c4", "d4", "e4", "f4", "g4", "h4",
"a5", "b5", "c5", "d5", "e5", "f5", "g5", "h5",
"a6", "b6", "c6", "d6", "e6", "f6", "g6", "h6",
"a7", "b7", "c7", "d7", "e7", "f7", "g7", "h7",
"a8", "b8", "c8", "d8", "e8", "f8", "g8", "h8",
]

altser_data=["0000000000000000000000000000000000000000000000000000000000000000000000000000000","000000000000000000000000000000000000000000000000000000000000000000000000000000000'"]

#time.sleep(3)

def Serial():
    try:
        ser = serial.Serial('/dev/ttyACM0', 9600)

    except serial.serialutil.SerialException:
        print("Bitte DGT-Brett anschließen")
        #lesSerial()

    else:
        #print("angeschlossen")
        try:
            ser_data = ser.readline()
        except serial.serialutil.SerialException:
            print("Bitte DGT-Brett anschließen")
        else:
            #print(ser_data)

            altser_data[1]=altser_data[0]
            altser_data[0]=ser_data
            Felder(ser_data, altser_data[1], mcount)


def checkGame():
    userdatajason= requests.get("https://lichess.org/api/user/"+user)
    userdata=userdatajason.text

    posP = userdata.find("playing")

    if posP<=1250:
        #print("Spielt gerade")

        id_game=userdata[posP+30:posP+38]
        #print(id_game)
        return(id_game)
    else:
        #print("Spielt gerade nicht")
        return("0")


def Felder(ser_data, altser_datan, mcount):
    if ser_data!=altser_datan:
        print("neues Spielfeld")
        #print(altser_data)

        #i=0
        for i in range(63):
            if ser_data[i]!=altser_datan[i]:
                if mcount==0:
                    ms=felder_name[i]
                    #print(ms)

                    halb_moves[1]=halb_moves[0]
                    halb_moves[0]=ms

                    if halb_moves[1]==halb_moves[0]:
                        mcount=0
                    else:
                        mcount=1


                elif mcount==1:
                    me=felder_name[i]
                    #print(me)

                    halb_moves[1]=halb_moves[0]
                    halb_moves[0]=me

                    if halb_moves[0]==halb_moves[1]:
                        mcount=1
                    else:
                        move=halb_moves[0]+halb_moves[1]

                        mcount=0
                        Move(move)


        i+=1


def Move(move):
    print("Zug ist: "+move)
    rf = input("Richtig/Fasch (y/n): ")
    if rf=="y":
        sendMove(move)

    elif rf=="n":
        print()
        print("Zürückstellen innerhalb 5 Sekunden und dann nochmal machen")
        time.sleep(5)
    else:
        print()
        print("Bitte Ja oder Nein eingeben (y/n)")
        Move(move)


def sendMove(move):
    print()

    game_id=checkGame()
    if game_id!="0":
        loop = asyncio.get_event_loop()
        loop.run_until_complete(main(move, game_id))
    else:
        print("Du bist gerade in keinem Spiel, bitte Figur zurückstellen(innerhalb 5 Sekunden) und ein Spiel starten")


async def main(move, id_game):
    client = APIClient(token=meintoken)
    response = await client.boards.make_move(game_id=id_game, move=move)
    #print(response)
    Lichess_response(response)


def Lichess_response(response):
    pos=find.response("content")
    posE=find.response("}", pos)
    zurück_lichess=response[pos+12:posE-1]
    pos=find.zurück_lichess("ok")

    if pos==-1:
        print(zurück_lichess)
    else:
        print("Zuck wurde erfolgreich ausgeführt")




while True:
    Serial()
    
    time.sleep(0.1)

main1.py
import requests
import asyncio
from lichess_client import APIClient
import serial
import time

user= "lucakreativ"
meintoken="QHwDl55O3vSbY9Ec"

mcount=0
halb_moves=["", ""]
#id_game="VKdMtBsg"
#moves="e2e4"

felder_name=[
"a1", "b1", "c1", "d1", "e1", "f1", "g1", "h1",
"a2", "b2", "c2", "d2", "e2", "f2", "g2", "h2",
"a3", "b3", "c3", "d3", "e3", "f3", "g3", "h3",
"a4", "b4", "c4", "d4", "e4", "f4", "g4", "h4",
"a5", "b5", "c5", "d5", "e5", "f5", "g5", "h5",
"a6", "b6", "c6", "d6", "e6", "f6", "g6", "h6",
"a7", "b7", "c7", "d7", "e7", "f7", "g7", "h7",
"a8", "b8", "c8", "d8", "e8", "f8", "g8", "h8",
]

altser_data=["x","x", "x"]


lesestat=0
#time.sleep(3)

def Serial(lesestat):
    try:
        ser = serial.Serial('/dev/ttyACM0', 9600)

    except serial.serialutil.SerialException:
        print("Bitte DGT-Brett anschließen")
        #lesSerial()

    else:
        #print("angeschlossen")
        try:
            ser_data = ser.readline()
        except serial.serialutil.SerialException:
            print("Bitte DGT-Brett anschließen")
        else:
            #print(ser_data)
            if len(ser_data)>=66:
                #print("Groß Genug")

                altser_data[2]=altser_data[1]
                altser_data[1]=altser_data[0]
                altser_data[0]=ser_data

                if altser_data[2]!= "x":
                    #print("nicht x")
                    Felder(ser_data, altser_data[1], mcount)
                else:
                    print("Keine Referenz")
            else:
                print("Datenübertragung ist zu kurz")


def checkGame():
    userdatajason= requests.get("https://lichess.org/api/user/"+user)
    userdata=userdatajason.text

    posP = userdata.find("playing")

    if posP<=1250:
        #print("Spielt gerade")

        id_game=userdata[posP+30:posP+38]
        #print(id_game)
        return(id_game)
    else:
        #print("Spielt gerade nicht")
        return("0")


def Felder(ser_data, altser_datan, mcount):
    if ser_data!=altser_datan:
        print("neues Spielfeld")
        #print(altser_data)

        #i=0
        for i in range(63):
            if ser_data[i]!=altser_datan[i]:
                if mcount==0:
                    ms=felder_name[i]
                    #print(ms)

                    halb_moves[1]=halb_moves[0]
                    halb_moves[0]=ms

                    if halb_moves[1]==halb_moves[0]:
                        mcount=0
                    else:
                        mcount=1


                elif mcount==1:
                    me=felder_name[i]
                    #print(me)

                    halb_moves[1]=halb_moves[0]
                    halb_moves[0]=me

                    if halb_moves[0]==halb_moves[1]:
                        mcount=1
                    else:
                        move=halb_moves[0]+halb_moves[1]

                        mcount=0
                        Move(move)


        i+=1


def Move(move):
    print("Zug ist: "+move)
    rf = input("Richtig/Fasch (y/n): ")
    if rf=="y":
        sendMove(move)
        print()
    elif rf=="n":
        print()
        print("Zürückstellen innerhalb 5 Sekunden und dann nochmal machen")
        time.sleep(5)
    else:
        print()
        print("Bitte Ja oder Nein eingeben (y/n)")
        Move(move)


def sendMove(move):
    print()

    game_id=checkGame()
    if game_id!="0":
        loop = asyncio.get_event_loop()
        loop.run_until_complete(main(move, game_id))
    else:
        print("Du bist gerade in keinem Spiel, bitte Figur zurückstellen(innerhalb 5 Sekunden) und ein Spiel starten")


async def main(move, id_game):
    client = APIClient(token=meintoken)
    response = await client.boards.make_move(game_id=id_game, move=move)
    #print(response)
    Lichess_response(response)


def Lichess_response(response):
    pos=find.response("content")
    posE=find.response("}", pos)
    zurück_lichess=response[pos+12:posE-1]
    pos=find.zurück_lichess("ok")

    if pos==-1:
        print(zurück_lichess)
    else:
        print("Zuck wurde erfolgreich ausgeführt")




while True:
    #lesestat=Serial(lesestat)
    Serial(lesestat)
    
    time.sleep(0.1)